proposals/type-classes.md

```diff
@@ -0,0 +1,296 @@
+
+```kotlin
+extension class Foo<A> {
+   val someKlazz = foo<A>() //won't compile because class disallow reified type args.
+}
+```
+
+```kotlin
+extension class Foo<A> {
+   val someKlazz = fooTC<A>() //works and does not requires to be inside an `inline reified` context.
+}
+```
+
+## Composition and chain of evidences
+
+Type class instances and declarations can encode further constrains in their generic args so they can be composed nicely:
+
+```kotlin
+package optionext
+
+extension class OptionMonoid<A>(with Monoid<A>): Monoid<Option<A>> {
+
+  val empty: Option<A> = None
+
+  fun Option.combine(ob: Option<A>): Option<A> =
+    when (this) {
+      is Some<A> -> when (ob) {
+                      is Some<A> -> Some(this.value.combine(b.value)) //works because there is evidence of a Monoid<A>
+                      is None -> ob
+                    }
+      is None -> this
+  }
+    
+}
+```
+
+The above instance declares a `Monoid<Option<A>>` as long as there is a `Monoid<A>` in scope.
+
+```kotlin
+import optionext.OptionMonoid
+import intext.IntMonoid
+
+Option(1).combine(Option(1)) // Option(2)
+Option("a").combine(Option("b")) // does not compile. Found `Monoid<Option<A>>` instance providing `combine` but no `Monoid<String>` instance was in scope
+```
+
+We believe the above proposed encoding fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other langs that also support typeclasses such as Scala where this is done via implicits.
+
+## Typeclasses over type constructors
+
+We recommend if this proposal is accepted that a lightweight version of higher kinds support is included to unveil the true power of typeclasses through the extensions mechanisms
+
+A syntax that would allow for higher kinds in these definitions may look like this:
+
+```kotlin
+extension interface FunctionK<F<_>, G<_>> {
+  fun <A> invoke(fa: F<A>): G<A>
+}
+
+extension object Option2List : FunctionK<Option, List> {
+  fun <A> invoke(fa: Option<A>): List<A> =
+    fa.fold({ emptyList() }, { listOf(it) })
+}
+``` 
+
+Here `F<_>` refers to a type constructor meaning a type that has a hole on it such as `Option`, `List`, etc. 
+
+A use of this declaration in a polymorphic function would look like:
+
+```kotlin
+fun <F<_>, A, B> transform(fa: F<A>, f: (A) -> B, with Functor<F>): F<B> = F.map(fa, f)
+
+transform(Option(1), { it + 1 }) // Option(2)
+transform("", { it + "b" }) // Does not compile: `String` is not type constructor with shape F<_>
+transform(listOf(1), { it + 1 }) // does not compile: No `Functor<List>` instance defined in scope.
+```
+
+## Language Changes
+
+- Add `with` to require instances evidences in both function and class/object declarations as demonstrated by previous and below examples:
```

Comment:
What about properties?

Reply:
Hi @ilya-g . If you are referring to flag properties as extension providers where in this example would be something like:
```kotlin
//somehere else in code
fun <A> evidence(with A): A = this

//inside a class or package
extension val optionMonoid: Monoid<Option<A>>
    get() = evidence<Monoid<Option<A>>()

//or with special sytnax

extension val optionMonoid: Monoid<Option<A>> with Monoid<Option<A>>
    get() = this
```
That could work but the issue with properties is that since they don't accept arguments we would need to provide a special way to encode the `with` constrains in the property itself and not sure how that would translate in java. If we add the constrain to an argument in the getter then it no longer has the `getter` shape expected by frameworks like java beans etc.

Reply:
This confuses me somehow. I can understand the first example, although my first impression is, that this is not how the extension mechanism should be applied. Is it a bad idea to use a local function here instead, and just limit the extension mechanism to functions?

The second example doesn't make sense to me^^ Could you give another example please? Although things are brought into scope, there should be a method call somewhere, shouldn't it?

Reply:
@hannespernpeintner there has been no discussion to supporting `extension` for properties. I'll be happy to limit to functions for now since `object` is already supported and can already provide a singleton value of the extension instance.

I realized after your comment the example was wrong, it should have looked like this:
```
extension val optionMonoid: Monoid<Option<A>> with Monoid<A>
    get() = evidence<Monoid<Option<A>>() //
```
That should be the correct form since to construct an `Monoid<Option<A>>` we need first evidence of `Monoid<A>`

Reply:
Thanks @Takhion, then we can probably support that use case too. Thanks for clarifying this.